File: Difficulty/Hard/[42] Trapping Rain Water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l, r, lower, level, water = 0, len(height)-1, 0, 0, 0
        while l < r:
            if height[r] > height[l]:
                lower = height[l]
                l += 1
            else:
                lower = height[r]
                r -= 1
            level = max(lower, level)
            water += level - lower
        return water

File: Difficulty/Hard/[42] Trapping Rain Water/solution.java
class Solution {
    public int trap(int[] height) {
        if (height.length == 0) return 0;
        int len = height.length;
        int left = 0, right = len - 1;
        int leftMax = height[0], rightMax = height[len-1];
        int res = 0;
        while (left <= right) {
            leftMax = Math.max(leftMax, height[left]);
            rightMax = Math.max(rightMax, height[right]);
            if (leftMax > rightMax) {
                res += rightMax - height[right];
                right--;
            } else {
                res += leftMax - height[left];
                left++;
            }
        }
        return res;
    }
}